File: runme/mcgen/local/datan/solsrc/s1gam.f
      PROGRAM S1GAM
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION XPL(1000),YPL(1000)
      CHARACTER*75 TX,TY
      PARAMETER(EPS=1.D-5)
C identify program to user
      WRITE(*,*)' Program S1GAM provides interactively graphs'
      WRITE(*,*)' of the gamma function and its inverse'
      WRITE(*,*)' '
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - gamma function'
      WRITE(*,*)' 2 - 1/gamma'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
      XMIN=-5.D0
      XMAX=5.D0
      YMIN=-5.D0
      YMAX=5.D0
      TX='x'
      LTX=1
      IF(ISWIT.EQ.1) THEN
        TY='&G@(x)'
        LTY=6
      ELSE IF(ISWIT.EQ.2) THEN
        TY='1/&G@(x)'
        LTY=8
      END IF
      CALL GROPEN
      CALL GROPWS(NWS)
      CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
      CALL GRVWWC(-0.15D0,0.9D0,0.15D0,0.85D0)
      CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
      CALL GRFRAM
      CALL GRBOUN
      CALL GRCCRS
      CALL GRSCLX(TX,LTX)
      CALL GRSCLY(TY,LTY)
      CALL GRSTCL(2)
      IF(ISWIT.EQ.1) THEN
C gamma
        XA=NINT(XMIN-.5D0)
5       XB=XA+1.D0
        IF(XA.LT.0.D0)THEN
C draw separate polylines between adjacent poles for x<0
          X1=XA+EPS
          X2=XB-EPS
          NPL=100
          DEL=(X2-X1)/DBLE(NPL-1)
          DO 10 I=1,NPL
            XPL(I)=X1+(I-1)*DEL
            YPL(I)=GGAMMA(XPL(I))
10        CONTINUE
          CALL GRPLIN(NPL,XPL,YPL)
          XA=XA+1.D0
          GO TO 5
        ELSE
C draw one polyline for x>0
          XA=XA+EPS
          NPL=1000
          DEL=(XMAX-XA)/DBLE(NPL-1)
          DO 20 I=1,NPL
            XPL(I)=XA+(I-1)*DEL
            YPL(I)=GGAMMA(XPL(I))
20        CONTINUE
          CALL GRPLIN(NPL,XPL,YPL)
        END IF
      ELSE IF(ISWIT.EQ.2) THEN
C 1/gamma
        NPL=1000
        DEL=(XMAX-XMIN)/DBLE(NPL-1)
        DO 30 I=1,NPL
          XPL(I)=XMIN+(I-1)*DEL
          YPL(I)=1.D0/GGAMMA(XPL(I))
30      CONTINUE
        CALL GRPLIN(NPL,XPL,YPL)
      END IF
      CALL GRCLSE
      END

File: runme/mcgen/local/datan/solsrc/s1gam.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2;
static double c_dmp15 = -.15, c_dp9 = .9, c_dp15 = .15, c_dp85 = .85,
              c_dmp414 = -.414, c_d1 = 1., c_d0 = 0.;

main(void)
{
    /* Local variables */
    static integer i, iswit, ltx, lty, nws, npl;
    static double del, x1, x2, xa, xb, xmin, ymin, xmax, ymax,
                  xpl[1000], ypl[1000];
    static char tx[75], ty[75];
    /* System generated locals */
    integer i__1;
    double d__1;
/* identify program to user */
    printf("%s\n", "Program S1GAM provides interactively graphs");
    printf("%s\n\n", "of the gamma function and its inverse.");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - gamma function");
    printf("%s\n", "2 - 1/gamma");
    printf("%s", "> ");
    scanf("%li", &iswit);
    (void)getchar();
    xmin = -5.;
    xmax = 5.;
    ymin = -5.;
    ymax = 5.;
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    strncpy(tx, "x", 1);
    ltx = 1;
    if (iswit == 1) {
        strncpy(ty, "&G@(x)", 6);
        lty = 6;
    } else if (iswit == 2) {
        strncpy(ty, "1/&G@(x)", 8);
        lty = 8;
    }
    gropen_();
    gropws_(&nws);
    grwncc_(&xmin, &xmax, &ymin, &ymax);
    grvwwc_(&c_dmp15, &c_dp9, &c_dp15, &c_dp85);
    grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
    grfram_();
    grboun_();
    grccrs_();
    grsclx_(tx, &ltx);
    grscly_(ty, &lty);
    grstcl_(&c__2);
    if (iswit == 1) {
/* gamma */
        d__1 = xmin - .5;
        xa = (double) nint(d__1);
L5:
        xb = xa + 1.;
        if (xa < 0.) {
/* draw separate polylines between adjacent poles for x<0 */
            x1 = xa + 1e-5;
            x2 = xb - 1e-5;
            npl = 100;
            del = (x2 - x1) / (double) (npl - 1);
            i__1 = npl;
            for (i = 1; i <= i__1; ++i) {
                xpl[i - 1] = x1 + (i - 1) * del;
                ypl[i - 1] = ggamma_(&xpl[i - 1]);
            }
            grplin_(&npl, xpl, ypl);
            xa += 1.;
            goto L5;
        } else {
/* draw one polyline for x>0 */
            xa += 1e-5;
            npl = 1000;
            del = (xmax - xa) / (double) (npl - 1);
            i__1 = npl;
            for (i = 1; i <= i__1; ++i) {
                xpl[i - 1] = xa + (i - 1) * del;
                ypl[i - 1] = ggamma_(&xpl[i - 1]);
            }
            grplin_(&npl, xpl, ypl);
        }
    } else if (iswit == 2) {
/* 1/gamma */
        npl = 1000;
        del = (xmax - xmin) / (double) (npl - 1);
        i__1 = npl;
        for (i = 1; i <= i__1; ++i) {
            xpl[i - 1] = xmin + (i - 1) * del;
            ypl[i - 1] = 1. / ggamma_(&xpl[i - 1]);
        }
        grplin_(&npl, xpl, ypl);
    }
    grclse_();
    return 0;
} /* main */

